### PR TITLE
[runtime env] Deletes the proto cache on RuntimeEnv

### PR DESCRIPTION
**Pull request by Catch-Bull**
## Why are these changes needed?

Mainly the following things:
- This PR deletes the proto cache on RuntimeEnv, ensuring that the user's modification of RuntimeEnv can take effect in the Proto message.
- validate whole runtime env when serialize runtime_env. 
- overload method `__setitem__` to parse and validate field when it has to modify.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by architkulkarni — APPROVED** [review]
Looks good!  One question, what's the benefit of validating every time `serialize()` is called?

File: python/ray/runtime_env.py [hunk 1]
```diff
@@ -367,58 +315,82 @@ def get_uris(self) -> List[str]:
...
+
     @classmethod
     def deserialize(cls, serialized_runtime_env: str) -> "RuntimeEnv":  # noqa: F821
         proto_runtime_env = json_format.Parse(serialized_runtime_env, ProtoRuntimeEnv())
         return cls.from_proto(proto_runtime_env)
 
     def serialize(self) -> str:
-
+        # To ensure the accuracy of Proto, `__setitem__` can only guarantee the
+        # accuracy of a certain field, not the overall accuracy
```
[the remark this answers is not in the document]
Comment (Catch-Bull) [reply python/ray/runtime_env.py:338]: yes,that's exactly what I mean

File: python/ray/serve/api.py [hunk 2]
```diff
@@ -485,9 +485,11 @@ def get_deploy_args(
 
         curr_job_env = ray.get_runtime_context().runtime_env
         if "runtime_env" in ray_actor_options:
-            ray_actor_options["runtime_env"].setdefault(
-                "working_dir", curr_job_env.get("working_dir")
-            )
+            # It is illegal to set field A to None.
```
[the remark this answers is not in the document]
Comment (Catch-Bull) [reply python/ray/serve/api.py:488]: oh... that's a typo actually... what I mean is set wokring_dir to None is illegal

[the remark this answers is not in the document]
Comment (Catch-Bull) [reply python/ray/serve/api.py]: I've been fixed it

File: python/ray/tests/test_runtime_env.py [hunk 3]
```diff
@@ -683,11 +686,182 @@ def test_serialize_deserialize(option):
...
+
+    # Test the interface related to working_dir
+    default_working_dir = "s3://bucket/key.zip"
+    modify_workeing_dir = "s3://bucket/key_A.zip"
+    runtime_env = RuntimeEnv(working_dir=default_working_dir)
+    runtime_env_dict = runtime_env.to_dict()
+    assert runtime_env.working_dir_uri() == default_working_dir
+    runtime_env["working_dir"] = modify_workeing_dir
+    runtime_env_dict["working_dir"] = modify_workeing_dir
+    assert runtime_env.working_dir_uri() == modify_workeing_dir
```
[the remark this answers is not in the document]
Comment (Catch-Bull) [reply python/ray/tests/test_runtime_env.py]: thanks!

File: python/ray/tests/test_runtime_env.py [hunk 4]
```diff
@@ -683,11 +686,182 @@ def test_serialize_deserialize(option):
...
+    runtime_env["container"].update(update_container)
+    runtime_env_dict["container"].update(update_container)
+    container_copy = container_init
+    container_copy.update(update_container)
+    assert runtime_env_dict == runtime_env.to_dict()
+    assert runtime_env.has_py_container()
+    assert runtime_env.py_container_image() == container_copy["image"]
+    assert runtime_env.py_container_worker_path() == container_copy["worker_path"]
+    assert runtime_env.py_container_run_options() == container_copy["run_options"]
+    # Test that the modification of concontainerda also works on
```
[the remark this answers is not in the document]
Comment (Catch-Bull) [reply python/ray/tests/test_runtime_env.py]: thanks, I've been fixed it.

**Comment by Catch-Bull** [comment]
> 
The main reason is to ensure that there is no problem with the data of runtime_env during serialization. The runtime_env_agent is not checked during deserialization. In the future, it may be checked on runtime_env_agent for multi-language purposes, but for now, I think we should check it on the python side

**Comment by Catch-Bull** [comment]
> Looks good! One question, what's the benefit of validating every time `serialize()` is called?

The main reason is to ensure that there is no problem with the data of runtime_env during serialization. The runtime_env_agent is not checked during deserialization. In the future, it may be checked on runtime_env_agent for multi-language purposes, but for now, I think we should check it on the python side

**Review by SongGuyang — APPROVED** [review]

**Pull request merged**